new combined orders mart table

KatelynGit opened the pull request:
# What are the relevant tickets?
https://github.com/mitodl/hq/issues/2425

# Description (What does it do?)
creates table marts__combined__orders from mitxonline, mixpro, and bootcamp data

# How can this be tested?
dbt build --select +marts__combined__orders

rachellougee commented on src/ol_dbt/models/marts/combined/marts__combined__orders.sql:
```diff
@@ -0,0 +1,65 @@
+        , product_id
+        , product_type
+        , user_email
+        , null as user_id
```
sorry I don't get why user_id would be null for xPro

KatelynGit replied:
i've populated it now

rachellougee commented on src/ol_dbt/models/marts/combined/marts__combined__orders.sql:
```diff
@@ -0,0 +1,65 @@
+        , order_created_on
+        , order_state
+        , null as courserun_id
+        , null as order_total_price_paid
```
xPro orders don't have total_price_paid? that seems strange

KatelynGit replied:
i've populated it now

rachellougee reviewed (changes requested):
I left a few comments, which I don't quite understand all the null for this combined orders. Most of these fields should have data populated.

KatelynGit commented on src/ol_dbt/models/marts/combined/_marts__combined__models.yml:
```diff
@@ -2,6 +2,36 @@
 version: 2
 
 models:
+- name: int__mitxpro__ecommerce_allorders
```
i've fixed this now

KatelynGit commented on src/ol_dbt/models/marts/combined/marts__combined__orders.sql:
```diff
@@ -0,0 +1,65 @@
+        , line_id
+        , order_created_on
+        , order_state
+        , null as courserun_id
```
i've populated it now

KatelynGit commented on src/ol_dbt/models/marts/combined/marts__combined__orders.sql:
```diff
@@ -0,0 +1,65 @@
+        , order_total_price_paid
+        , null as product_id
+        , null as product_type
+        , null as user_email
```
i've populated it now

KatelynGit commented:
> I left a few comments, which I don't quite understand all the null for this combined orders. Most of these fields should have data populated.

Thanks @rachellougee! I've made changes per your comments

rachellougee reviewed (commented):
Beside these comments, i also can't run this mart as it appears some issues with schema naming conflicts

@KatelynGit were you able to run this new mart without issue?

KatelynGit commented on src/ol_dbt/models/marts/combined/marts__combined_orders.sql:
```diff
@@ -0,0 +1,114 @@
+    left join mitxpro__ecommerce_order
+        on mitxpro__ecommerce_allorders.order_id = mitxpro__ecommerce_order.order_id
+    left join mitxpro__b2becommerce_b2border
+        on mitxpro__ecommerce_allorders.b2border_id = mitxpro__b2becommerce_b2border.b2border_id
```
the problem is some of the fields aren't there in allorders so I needed to use those other tables for now

rachellougee replied:
I am not sure if I understand your comment, my point was that this join and additional table import can be totally eliminated if `order_purchaser_user_id`, `order_total_price_paid`, `b2border_total_price` are added in int__mitxpro__ecommerce_allorders since it's already [joined all these tables](https://github.com/mitodl/ol-data-platform/blob/main/src/ol_dbt/models/intermediate/mitxpro/int__mitxpro__ecommerce_allorders.sql#L62-L99). Also, these fields exist in MITx Online intermediate orders table.

That's not to say we can't do join in mart as there are many cases we will need additional joins and aggregation in mart. Unfortunately I just don't think this is not one of cases. With that said, it's up to you.

KatelynGit commented:
> Beside these comments, i also can't run this mart as it appears some issues with schema and CTE name
> 
> @KatelynGit were you able to run this new mart without issue?

@rachellougee  sorry about that - can you try again. I believe it should work now

KatelynGit commented on src/ol_dbt/models/marts/combined/_marts__combined__models.yml:
```diff
@@ -2,6 +2,36 @@
+  - name: user_email
+    description: str, user email associated with their account
+  - name: user_id
+    description: int, foreign key to the user table
```
I've added a unique test as well

rachellougee reviewed (commented):
@KatelynGit I am getting error on the duplicates for xpro. It looks like some of these duplicates are from using different coupon.

```16:57:41  Failure in test dbt_expectations_expect_compound_columns_to_be_unique_marts__combined__orders_order_id__line_id__b2b_indicator__platform (models/marts/combined/_marts__combined__models.yml)
16:57:41    Got 292 results, configured to fail if >10
```

Can you double check these records and run a build after making a change? Additionally I don't know these fields are enough to be in combined mart for orders since it doesn't have coupon, receipt/payment data in it. But that can be another subject.

KatelynGit commented:
> @KatelynGit I am getting error on the duplicates for xpro. It looks like some of these duplicates are from using different coupon.
> 
> ```
> 16:57:41  Failure in test dbt_expectations_expect_compound_columns_to_be_unique_marts__combined__orders_order_id__line_id__b2b_indicator__platform (models/marts/combined/_marts__combined__models.yml)
> 16:57:41    Got 292 results, configured to fail if >10
> ```
> 
> Can you double check these records and run a build when making a change? Additionally I don't know these fields are enough to be in combined mart for orders since it doesn't have coupon, receipt/payment data in it. But that can be another subject.

I've added coupon_id and name to the table and added them to the unique constraint it seems to have fixed the issue

rachellougee reviewed (approved):
The unique test runs fine now
